### ISSUE TITLE
Reduce generated bundle size

### ISSUE THREAD

**Issue by jooola** [directus/app#2575]
I'd like to make the app a bit faster. The first thing I proposed is to lazyload the components in the router (see #2573). Now I'd like to reduce the bundle size.

Using `webpack-bundle-analyzer` I generated the following schema:
DEV:
![image](https://user-images.githubusercontent.com/19195485/73991701-42306a00-494d-11ea-8287-77d52be6e3c3.png)
PROD:
![image](https://user-images.githubusercontent.com/19195485/73991633-0bf2ea80-494d-11ea-8a97-cb958e6ffaa7.png)

## slug

The first big file is `./node_modules/unicode/category/So.js`. It is a dependency of the `slug` package: https://www.npmjs.com/package/slug

The first question is: Do we really need to support Unicode for the 4 function calls in the project ?
![image](https://user-images.githubusercontent.com/19195485/73991654-16ad7f80-494d-11ea-9bcb-4287cf64abd3.png)

## TinyMCE

TinyMCE also has a nice size, I would recommend to lazy load this dependency from the wysiwyg interface component, so that the library is only loaded when we need it and also split in a separate file.

Maybe this could be done for all the interfaces that rely on external dependencies, like leaflet, codemirror, ...

## lodash

Then we have lodash `/node_modules/lodash/lodash.js`. It is a dependency of the `slug` package: https://www.npmjs.com/package/slug

It seems every lodash functions are imported individually, but we need to find out whether the entire lib is bundled. https://stackoverflow.com/questions/43479464/how-to-import-a-single-lodash-function

**Comment by rijkvanzanten** [comment]
> I'd like to make the app a bit faster. 

The total download size of a fresh open of Directus currently with caches turned off is about 3.5 MB. While it's always a good thing to reduce the amount of code shipped in the initial request, I would to point out that there are way more speed improvements to gain in the API side of things that will drastically improve the perceived performance of the application. 

I would also like to point out that the initial load time is way less important in a purely back-end admin focused app like this than the load times while _using_ the app. For example, waiting an extra second or two while opening the app for the first time is wait less annoying than having to wait to watch interfaces and other UI elements pop into place sporadically. 

While I'm totally down to decrease overall bundle size by optimizing and getting rid of dependencies we might not fully need, I'm a bit hesitant to go down the rabbithole that is micro-optimizations in load order. In Directus, I'm currently more in camp "load everything up front, have the app be super stable and fast after" than camp "make initial load as fast as possible" (@benhaynes)

**Comment by benhaynes** [comment]
Thanks for the research and feedback, @jooola — very helpful assessment! We're all very excited to optimize the app in general. ❤️

On these specific changes, my opinion is that the priority should be getting the App as fast and responsive as possible during usage... even if at the sight expense of upfront load time. (as Rijk said).

With that key goal in mind, let's see what optimiations are best.

**Comment by rijkvanzanten** [comment]
Also, the graphs above are a little misleading. The only scripts downloaded on first load are:

```
script.js
chunk-vendors
app
```

(script.js is the user custom JS file, empty by default)

<img width="884" alt="image" src="https://user-images.githubusercontent.com/9141017/73993838-6ece8c80-4921-11ea-8c2f-435343b39a49.png">

which together are not nearly half a MB gzipped.

All other little chunks are loaded asynchronously already. Realistically speaking, on initial load on fresh caches, the graph looks more like:

![Untitled](https://user-images.githubusercontent.com/9141017/73993959-cec53300-4921-11ea-95db-ead5478fab7a.png)

---

That all taken into account, we _do_ see (like you correctly mentioned above) that that Unicode SO file is _gigantic_. Swapping out `slug` sounds like a very good first step.

> Do we really need to support Unicode for the 4 function calls in the project ?

Yes; someone raised an issue before where things like ß and ü weren't properly handled in the slug interface. We have to keep in mind that Directus is used by an international audience, and the slug interface is a very often used interface, so it should gracefully handle whatever people throw at it. That being said, I couldn't care less about the slug interface automatically transform things like `<` to `less-than` or `♲` to `recycling`

**Comment by rijkvanzanten** [comment]
Another also: I really appreciate your enthusiasm and eagerness to jump right [into PRs](https://github.com/directus/app/pull/2571) 🙂 Apologies if it felt like I shot down your effort too harshly. Just wanted to make sure you wouldn't spend hours cleaning up legacy code only to learn that things were done certain ways on purpose!

**Comment by benhaynes** [comment]
Another thought I had (maybe worth discussing) is: could we start start asynchronously loading all the JS, CSS, and other files, directly after the login page initially loads?

I feel like there will always be at least three to five seconds after the login page loads but before the user finishes authenticating. Isn't that a non-obtrusive time we could asnc load the rest of the data?

Is it wrong in some way to load data before someone commits to logging in? 🤔

**Comment by rijkvanzanten** [comment]
Yeah! That's what I meant by https://github.com/directus/app/issues/2495 too :)

**Comment by jooola** [comment]
> Yes; someone raised an issue before where things like ß and ü weren't properly handled in the slug interface (this was a while ago in v6). We have to keep in mind that Directus is used by an international audience, and the slug interface is a very often used interface, so it should gracefully handle whatever people throw at it. That being said, I couldn't care less about the slug interface automatically transform things like < to less-than or ♲ to recycling

The end of your comment just suggest that we might use another library that don't contains all the fancy emojis. This needs some more research but a replacement could be https://www.npmjs.com/package/slugify

> Another also: I really appreciate your enthusiasm and eagerness to jump right into PRs slightly_smiling_face Apologies if it felt like I shot down your effort too harshly. Just wanted to make sure you wouldn't spend hours cleaning up legacy code only to learn that things were done certain ways on purpose!

Don't apologize, it is my mistake not to raise an issue first and directly dive into the code, it a common mistake that needs to be reminded time to time. I simply felt like coding so I don't raised any issue/questions.

Agree on the rest of your answerss, at least this issue tells us the current state of the bundle size which is always good.

**Comment by rijkvanzanten** [comment]
From a first (arguably dumb / too limited) test, `slugify` seems to cover the two basics that were a problem before:

<img width="561" alt="image" src="https://user-images.githubusercontent.com/9141017/74037777-f998a100-498c-11ea-876e-aca4fb7a2cb0.png">

Based on bundlephobia.com, moving from `slug` to `slugify` should decrease our initial load bundle size by `102.4kB` (gzipped). That sounds like an easy win @jooola

**Comment by jooola** [comment]
> Based on bundlephobia.com, moving from slug to slugify should decrease our initial load bundle size by 102.4kB (gzipped). That sounds like an easy win @jooola

Haha, yeah, small win. Not sure it compensate the loose of certainty/stability regarding slug creation.

I don't think I have anything to add on this matter, has been a nice try.

**Comment by everyx** [comment]
@rijkvanzanten @jooola does `slugify` support CJK string?
